### ISSUE TITLE
feature: add algorithm.sortPartial (eg quicksort that stops early)

### ISSUE THREAD

**Issue by timotheecour** [timotheecour/Nim#390]

**Comment by ringabout** [comment]
If we choose heap based implementation([cpp](https://en.cppreference.com/w/cpp/algorithm/partial_sort)), `heapqueue` needs support user-defined cmp function.

**Comment by timotheecour** [comment]
> heapqueue needs support user-defined cmp function.

that's needed no matter what.

> If we choose heap based implementation

we may need both, because for top K with an array of size N, if K is small, heap based sort will be fastest whereas for large K (eg N/2) quicksort (or similar) will be fastest.

**Comment by ringabout** [comment]
Yeah, nlargest(Python) combines heapq and tim-sort to get the best performance.
